src/LanguageServer/Protocol/Handler/PullHandlers/VersionedPullCache.CacheItem.cs

```diff
@@ -0,0 +1,139 @@
+{
+    /// <summary>
+    /// Internal cache item that updates state for a particular <see cref="Workspace"/> and <see cref="ProjectOrDocumentId"/> in <see cref="VersionedPullCache{TCheapVersion, TExpensiveVersion, TState, TComputedData}"/>
+    /// This type ensures that the state for a particular key is never updated concurrently for the same key (but different key states can be concurrent).
+    /// </summary>
+    private class CacheItem(string uniqueKey)
+    {
+        /// <summary>
+        /// Guards access to <see cref="_lastResult"/>.
+        /// This ensures that a cache entry is fully updated in a single transaction.
+        /// </summary>
+        private readonly SemaphoreSlim _semaphore = new(1);
+
+        /// <summary>
+        /// Stores the current state associated with this cache item.
+        /// 
+        /// <list type="bullet">
+        ///   <item>The resultId reported to the client.</item>
+        ///   <item>The TCheapVersion of the data that was used to calculate results.
+        ///       <para>
+        ///       Note that this version can change even when nothing has actually changed (for example, forking the 
+        ///       LSP text, reloading the same project). So we additionally store:</para></item>
+        ///   <item>A TExpensiveVersion (normally a checksum) checksum that will still allow us to reuse data even when
+        ///   unimportant changes happen that trigger the cheap version change detection.</item>
+        ///   <item>The hash of the data that was computed when the resultId was generated.
+        ///       <para>
+        ///       When the versions above change, we must recalculate the data.  However sometimes that data ends up being exactly the same as the prior request.
+        ///       When that happens, this allows us to send back an unchanged result instead of reserializing data the client already has.
+        ///       </para>
+        ///   </item>
+        /// </list>
+        /// 
+        /// </summary>
+        private (string resultId, TCheapVersion cheapVersion, TExpensiveVersion expensiveVersion, int hashedData)? _lastResult;
+
+        /// <summary>
+        /// Updates the values for this cache entry.  Guarded by <see cref="_semaphore"/>
+        /// 
+        /// Returns null if the previousPullResult can be re-used, otherwise returns a new resultId and the new data associated with it.
+        /// </summary>
+        public async Task<(string, ImmutableArray<TComputedData>)?> UpdateCacheItemAsync(
+            VersionedPullCache<TCheapVersion, TExpensiveVersion, TState, TComputedData> cache,
+            PreviousPullResult? previousPullResult,
+            bool isFullyLoaded,
+            TState state,
+            CancellationToken cancellationToken)
+        {
+            // Ensure that we only update the cache item one at a time.
+            // This means that the computation of new data for this item only occurs sequentially.
```

Comment:
aside: have we considered computing results in parallel *higher up* in workspace diagnostics?  we walk the diagnostic sources serially.  but we could use our parallel helpers to potentially speed that up.

Reply:
Definitely potentially possible.  I'll leave that to investigate in a different PR.